message-editing: include stream icon in message sent  and visibility notification banners

wandrew0 opened the pull request:
<!-- Describe your pull request here.-->
Fix the "Compose banners" item in #25257. Compose banners include message validation and message notification banners, created in compose_validate.ts and compose_notification.ts. This PR puts a stream icon in these banners.
    

Fixes: #25257 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
**_message notifications and links_**
![Screenshot from 2024-04-28 14-08-53](https://github.com/zulip/zulip/assets/73965466/5112b687-0970-4b71-bb71-4f10b1cded62)
![Screenshot from 2024-04-28 14-09-57](https://github.com/zulip/zulip/assets/73965466/4fe6439e-dfd4-49bb-ae94-843b72d739f5)
![Screenshot from 2024-04-28 14-13-06](https://github.com/zulip/zulip/assets/73965466/69c47144-0d8d-4933-b005-fab60f27af21)
![Screenshot from 2024-04-28 14-13-21](https://github.com/zulip/zulip/assets/73965466/c0829dfd-0f56-42d0-9473-8a0d4465ecac)
![Screenshot from 2024-04-28 14-14-14](https://github.com/zulip/zulip/assets/73965466/d485f362-3235-4867-87e4-63a02feb7853)
![Screenshot from 2024-04-28 14-14-32](https://github.com/zulip/zulip/assets/73965466/853c0ad4-b362-4236-8e08-82d874f8b2db)

**_message validation warnings_**

![Screenshot from 2024-04-28 14-51-22](https://github.com/zulip/zulip/assets/73965466/ac304231-9dba-4e0f-bc2e-1c4b85ab18a3)
![Screenshot from 2024-04-28 14-24-19](https://github.com/zulip/zulip/assets/73965466/8476e699-b998-430a-8ddd-2f5539cb9d79)
![Screenshot from 2024-04-28 14-58-01](https://github.com/zulip/zulip/assets/73965466/f340284e-2422-4b54-8473-76b217106e22)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks for the fix! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?

alya commented:
Please also complete the self-review checklist (as applicable), and post a comment to request a review when ready.

wandrew0 commented:
> Please also complete the self-review checklist (as applicable), and post a comment to request a review when ready.

@alya I completed the checklist and updated the commit message. (I will install commit hooks in my other PRs and fix them). It's ready for review. Thank you!

alya commented:
Thanks! The horizontal spacing in your screenshots looks quite a bit off to me.

What's the relationship between this PR and #20562, where we previously worked on this?

wandrew0 commented:
@alya see my replies below in line. Thanks!
> Thanks! The horizontal spacing in your screenshots looks quite a bit off to me.
I reused the render_inline_decorated_stream_name function(inline_decorated_stream_name.hbs file) to generate stream names with icons, which is why the spacing between icon and stream name is. Do you want me to create another hbs file to generate stream names for banners?
> 
> What's the relationship between this PR and #20562, where we previously worked on this?
#20562 tries to do the same thing, except the code it's based on is too old for me to do a merge instead. #20562 creates 3 CSS styles for streams, each for a different icon, and passes in the three parameters (isPrivate, isGlobal, isNormal); depending on which is true, it picks one of the stream icon CSS styles to use. 
Personally, I think the current render_inline_decorated_stream_name is better because we have a centralized place to generate icons and streams.

alya commented:
> Would you like me to create another HBS file to generate banner stream names?

I have no idea what implementation is appropriate; you can start a discussion in #frontend to get feedback. My point is just that having the privacy icon closer to the word on the left than to the stream name does not look good, and I'm pretty sure we don't have this issue in some other banners.

wandrew0 commented:
> > Would you like me to create another HBS file to generate banner stream names?
> 
> I have no idea what implementation is appropriate; you can start a discussion in #frontend to get feedback. My point is just that having the privacy icon closer to the word on the left than to the stream name does not look good, and I'm pretty sure we don't have this issue in some other banners.

Yes, I agree. I will make appropriate changes to make the icons closer to the stream name. Thanks.

wandrew0 commented:
@alya I have fixed the spacing issue by adding a 3px right shift, so the icon is not exactly in the center. I have updated the screenshot. It's ready for review. Thanks.

alya commented:
If we move forward with #29186, the  "Your message is outside your current view." banner will no longer exist, so it's probably not worth spending time on that one. Do take a look at the other banners noted in https://github.com/zulip/zulip/issues/25257, though.

wandrew0 commented:
> If we move forward with #29186, the "Your message is outside your current view." banner will no longer exist, so it's probably not worth spending time on that one. Do take a look at the other banners noted in #25257, though.

@alya, okay, I won't fix the "Your message is outside your current view" link issue. I have fixed the "Now following.." link; see the attached screenshots. I will look into #25257 and put stream icons in other places. Let me know of any issues. Thank you.

wandrew0 commented:
@alya It's ready for review. I have fixed the message compose banners, i.e., the "Compose Banner" item in #25257. I have also cleaned up the commit and uploaded relevant screenshots. I will continue to look into other items in #25257, as this PR has many screenshots, I will submit separate PR for other items. Thank you.

alya commented:
Hm, now the icons look weirdly high; I'm not convinced moving them up was the right fix for the underline weirdness.

Rather than me continuing to review this based solely on the visuals, it's probably best for someone to take a look at the code at this point. @prakhar1144 perhaps you could take a look?

karlstolley commented:
Getting the icon alignment right here is going to take some doing. Containers like `.above_compose_banner_action_link` should probably be set to `display: inline-flex`, but then there is other stuff going on between box heights being smaller than line-heights, and icons having fractional heights, e.g., 11.25px adjacent a fractional line-height. My thought is that that work will probably be complex enough that it'll deserve its own PR, so close-enough is probably the name of the game here.

chandanSahoo-cs commented:
Can you assign this issue to me. If it still has issues

Muppasanipraneeth commented:
@zulipbot  if not assing assing it to me

SinghArmaanJ commented:
@zulipbot claim

Md-Murthaza commented:
@timabbott  this issue is been fixed and need to merge.but still its open

swarnalicoder commented:
Hi @wandrew0 and team,

I’m interested in contributing to the stream icon feature in compose and notification banners. Could you please guide me on where I can help or any specific tasks that need attention?

Looking forward to contributing!

khushiiii24 commented:
@zulipbot claim

khushiiii24 commented:
@zulipbot unassign

devpaths commented:
@zulipbot claim

rialdcart098 commented:
is this still open?

user-abhay08 commented:
@zulipbot claim
